frontend/src/components/ide/ProgramIde.tsx

```diff
@@ -217,6 +242,10 @@ const ProgramIde: React.FC<{
         />
         <StackInfo className={localClasses.stackInfo} />
         <CodeEditor className={localClasses.editor} />
+        <Prompt
+          when={sourceCode !== userProgram.sourceCode}
+          message="You have unsaved changes. Are you sure you want to leave?"
+        />
```

Comment:
This is all the logic for the unsaved changes prompt. React router handles the rest.